### PR TITLE
Ion tag

### PR DESCRIPTION
**Pull request by lucasrasec**
Implementing the Tag component, with tests and documentation  #82

### CODE REVIEW

**Review by Deeved — APPROVED** [review]

File: projects/ion/src/lib/core/types/context.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export type Context = 'success' | 'info' | 'warning' | 'negative' | 'neutral';
```
Comment (iurynogueira) [inline projects/ion/src/lib/core/types/context.ts:1]: You can add this in status.ts like TagStatusType;

export type TagStatusType = 'success' | 'info' | 'warning' | 'negative' | 'neutral';

File: projects/ion/src/lib/tag/tag.component.html [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<div
+  data-testid="ion-tag"
+  [style.color]="!context && validateColor()"
+  [style.background]="!context && validateColor() + '1A'"
+  [class]="setTagType()"
+>
+  <ion-icon
+    data-testid="tag-icon"
+    class="tag-icon"
+    size="15"
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.html:10]: Icon size is 12.

image

File: projects/ion/src/lib/tag/tag.component.scss [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+.ion-tag {
+  display: flex;
+  align-items: center;
+  flex-direction: row;
+  border-radius: 50px;
+  padding: 6px spacing(1.5);
+  justify-content: center;
+  align-items: center;
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.scss:22]:
```suggestion
```

repeated

File: projects/ion/src/lib/tag/tag.component.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  setTagType() {
+    return `ion-tag ${this.outline ? 'outline' : ''} ${
+      this.context ? this.context : ''
+    }`;
+  }
+
+  validateColor() {
+    return /^#([0-f]{6})$/.test(this.color) ? this.color : '#505566';
+  }
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.ts:33]:
```suggestion
  validateHexColor() {
    return /^#([0-f]{6})$/.test(this.color) ? this.color : '#505566';
  }
```
I like it, you can add this in utils?

File: projects/ion/src/lib/tag/tag.component.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+@Component({
+  selector: 'ion-tag',
+  templateUrl: './tag.component.html',
+  styleUrls: ['./tag.component.scss'],
+})
+export class TagComponent {
+  @Input() public outline = true;
+  @Input() public label!: string;
+  @Input() public context?: Context;
+  @Input() public color?: string = '#505566';
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.ts:22]:
```suggestion
  @Input() public color?: string = this.defaultColor;
```

File: projects/ion/src/lib/tag/tag.component.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  @Input() public icon?: IconType = 'trash';
+
+  setTagType() {
+    return `ion-tag ${this.outline ? 'outline' : ''} ${
+      this.context ? this.context : ''
+    }`;
+  }
+
+  validateColor() {
+    return /^#([0-f]{6})$/.test(this.color) ? this.color : '#505566';
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.ts:32]:
```suggestion
    return /^#([0-f]{6})$/.test(this.color) ? this.color : this.defaultColor;
```

File: projects/ion/src/lib/tag/tag.component.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  icon?: string;
+}
+
+@Component({
+  selector: 'ion-tag',
+  templateUrl: './tag.component.html',
+  styleUrls: ['./tag.component.scss'],
+})
+export class TagComponent {
+  @Input() public outline = true;
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.ts:19]:
```suggestion
  private defaultColor = '#505566';
  @Input() public outline = true;
```

File: projects/ion/src/lib/tag/tag.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  it('should render default message', async () => {
+    await sut();
+    expect(screen.getByText(defaultValue.label)).toBeInTheDocument();
+  });
+
+  it('should render with default color', async () => {
+    await sut();
+    expect(screen.getByTestId(IDs.tag)).toHaveStyle(`color: ${defaultColor};`);
+  });
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.spec.ts:55]: Create beforeEach

File: projects/ion/src/lib/tag/tag.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+describe('TagComponent', () => {
+  it('should render component', async () => {
+    await sut();
+    expect(screen.getByTestId(IDs.tag)).toBeInTheDocument();
+  });
+
+  it('should render with border', async () => {
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.spec.ts:42]:
```suggestion
  it('should render with border by default', async () => {
```

File: projects/ion/src/lib/tag/tag.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+  it.each(tagTypes)('should render %s tag', async (type) => {
+    await sut({ ...defaultValue, context: type });
+    expect(screen.getByTestId(IDs.tag)).toHaveClass(type);
+  });
+
+  it('should render icon', async () => {
+    await sut();
+    expect(screen.getByTestId(IDs.icon)).toBeInTheDocument();
+  });
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.spec.ts:76]: add in same describe with beforeEach

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: projects/ion/src/lib/tag/tag.component.html (hunk 2 above)
Comment (lucasrasec) [reply projects/ion/src/lib/tag/tag.component.html:10]: In the implementation leaving the icon at 12px made it to small
12px
image
15px
image

should i put 12 or 15?

File: projects/ion/src/lib/tag/tag.component.html [hunk 11]
```diff
@@ -0,0 +1,15 @@
+<div
+  data-testid="ion-tag"
+  [style.color]="!status && validateHexColor()"
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.html:3]:
```suggestion
  [style.color]="!status && validateHexColor()"
```

This validation is repeated 3 times here, what do you think in create a single function to this?

File: projects/ion/src/lib/tag/tag.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    async (color) => {
+      await sut({ ...defaultValue, color: color });
+      expect(await screen.findByTestId(IDs.tag)).toHaveStyle(
+        `background: ${color}1A;`
+      );
+    }
+  );
+
+  it.each(tagTypes)('should render %s tag even with a color', async (type) => {
+    await sut({ ...defaultValue, status: type, color: '#be531c' });
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.spec.ts:102]:
```suggestion
    const color = '#be531c';
    await sut({ ...defaultValue, status: type, color });
```

File: projects/ion/src/lib/tag/tag.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    expect(screen.getByTestId(IDs.tag)).toHaveClass(type);
+    expect(await screen.findByTestId(IDs.tag)).not.toHaveStyle(
+      'color: #be531c;'
```
Comment (iurynogueira) [inline projects/ion/src/lib/tag/tag.component.spec.ts:105]:
```suggestion
      `color: ${color};`
```

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: projects/ion/src/lib/tag/tag.component.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+export interface IonTagProps {
+  outline?: boolean;
+  status?: TagStatus;
+  color?: string;
+  label: string;
+  icon?: string;
+}
+
+const defaultColor = '#505566';
```
Comment (edsonsnunes) [inline projects/ion/src/lib/tag/tag.component.ts:14]: Não teria que colocar $neutral-7?

Comment (lucasrasec) [reply projects/ion/src/lib/tag/tag.component.ts:14]: Tem como utilizar as variáveis do colors em um arquivo .ts? Acho que eles são só para os .scss não?

**Review by edsonsnunes — APPROVED** [review]

**Review by iurynogueira — APPROVED** [review]

**Pull request merged**